<a href="https://colab.research.google.com/github/marreapato/Deep_Learning_Course/blob/main/PoliticsExercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Keras Exercise

## Predict political party based on votes

As a fun little example, we'll use a public data set of how US congressmen voted on 17 different issues in the year 1984. Let's see if we can figure out their political party based on their votes alone, using a deep neural network!

For those outside the United States, our two main political parties are "Democrat" and "Republican." In modern times they represent progressive and conservative ideologies, respectively.

Politics in 1984 weren't quite as polarized as they are today, but you should still be able to get over 90% accuracy without much trouble.

Since the point of this exercise is implementing neural networks in Keras, I'll help you to load and prepare the data.

Let's start by importing the raw CSV file using Pandas, and make a DataFrame out of it with nice column labels:

In [1]:
from google.colab import drive
drive.mount('/content/gdrive', force_remount=True)

Mounted at /content/gdrive


In [2]:



%cd /content/gdrive/MyDrive/mlcrs

!ls



/content/gdrive/MyDrive/mlcrs
access_log.txt			      KFoldCrossValidation.ipynb     Percentiles.ipynb
breakfast.jpg			      KMeans.ipynb		     PoliticsExercise.ipynb
bridge.jpg			      LinearRegression.ipynb	     PolynomialRegression.ipynb
bunny.jpg			      mammographic_masses.data.txt   Python101.ipynb
castle.jpg			      mammographic_masses.names.txt  regression.txt
ConditionalProbabilityExercise.ipynb  mammo_masses_project.ipynb     Seaborn.ipynb
ConditionalProbabilitySolution.ipynb  MatPlotLib.ipynb		     SparkDecisionTree.py
CovarianceCorrelation.ipynb	      MeanMedianExercise.ipynb	     SparkKMeans.py
DecisionTree.ipynb		      MeanMedianMode.ipynb	     SparkLinearRegression.py
DeepLearningProject.ipynb	      ml-100k			     SparkPCA.py
DeepLearningProject-Solution.ipynb    MLCourse.zip		     StdDevVariance.ipynb
distancerecomendation-KNN.ipynb       mlcrsItemBasedCF.ipynb	     subset-small.tsv
Distributions.ipynb		      mlcrsQ-Learning.ipynb	     SVC.ipynb
emails				      mlcrs_Sim

In [3]:
data_dir = "/content/gdrive/MyDrive/mlcrs/house-votes-84.data.txt"
data_dir2 = "/content/gdrive/MyDrive/Machine Learning e Data Science com Python de A à Z/Bases de dados/credit.pkl"

In [4]:
import pandas as pd

feature_names =  ['party','handicapped-infants', 'water-project-cost-sharing',
                    'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                    'el-salvador-aid', 'religious-groups-in-schools',
                    'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                    'mx-missle', 'immigration', 'synfuels-corporation-cutback',
                    'education-spending', 'superfund-right-to-sue', 'crime',
                    'duty-free-exports', 'export-administration-act-south-africa']

voting_data = pd.read_csv(data_dir, na_values=['?'],
                          names = feature_names)
voting_data.head()

party handicapped-infants water-project-cost-sharing  \
0  republican                   n                          y   
1  republican                   n                          y   
2    democrat                 NaN                          y   
3    democrat                   n                          y   
4    democrat                   y                          y   

  adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
0                                 n                    y               y   
1                                 n                    y               y   
2                                 y                  NaN               y   
3                                 y                    n             NaN   
4                                 y                    n               y   

  religious-groups-in-schools anti-satellite-test-ban  \
0                           y                       n   
1                           y                       n   
2                           y                       n   
3                           y                       n   
4                           y                       n   

  aid-to-nicaraguan-contras mx-missle immigration  \
0                         n         n           y   
1                         n         n           n   
2                         n         n           n   
3                         n         n           n   
4                         n         n           n   

  synfuels-corporation-cutback education-spending superfund-right-to-sue  \
0                          NaN                  y                      y   
1                            n                  y                      y   
2                            y                  n                      y   
3                            y                  n                      y   
4                            y                NaN                      y   

  crime duty-free-exports export-administration-act-south-africa  
0     y                 n                                      y  
1     y                 n                                    NaN  
2     y                 n                                      n  
3     n                 n                                      y  
4     y                 y                                      y

We can use describe() to get a feel of how the data looks in aggregate:

In [5]:
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        435                 423                        387   
unique         2                   2                          2   
top     democrat                   n                          y   
freq         267                 236                        195   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                424                  424             420   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 253                  247             212   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          424                     421   
unique                           2                       2   
top                              y                       y   
freq                           272                     239   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        420       413         428   
unique                         2         2           2   
top                            y         y           y   
freq                         242       207         216   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           414                404                    410   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            264                233                    209   

       crime duty-free-exports export-administration-act-south-africa  
count    418               407                                    331  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     248               233                                    269

We can see there's some missing data to deal with here; some politicians abstained on some votes, or just weren't present when the vote was taken. We will just drop the rows with missing data to keep it simple, but in practice you'd want to first make sure that doing so didn't introduce any sort of bias into your analysis (if one party abstains more than another, that could be problematic for example.)

In [6]:
voting_data.dropna(inplace=True)
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        232                 232                        232   
unique         2                   2                          2   
top     democrat                   n                          n   
freq         124                 136                        125   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                232                  232             232   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 123                  119             128   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          232                     232   
unique                           2                       2   
top                              y                       y   
freq                           149                     124   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        232       232         232   
unique                         2         2           2   
top                            y         n           y   
freq                         119       119         128   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           232                232                    232   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            152                124                    127   

       crime duty-free-exports export-administration-act-south-africa  
count    232               232                                    232  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     149               146                                    189

Our neural network needs normalized numbers, not strings, to work. So let's replace all the y's and n's with 1's and 0's, and represent the parties as 1's and 0's as well.

In [7]:
voting_data.replace(('y', 'n'), (1, 0), inplace=True)
voting_data.replace(('democrat', 'republican'), (1, 0), inplace=True)

In [8]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

Finally let's extract the features and labels in the form that Keras will expect:

In [9]:
all_features = voting_data[feature_names].drop('party', axis=1).values
all_classes = voting_data['party'].values

OK, so have a go at it! You'll want to refer back to the slide on using Keras with binary classification - there are only two parties, so this is a binary problem. This also saves us the hassle of representing classes with "one-hot" format like we had to do with MNIST; our output is just a single 0 or 1 value.

Also refer to the scikit_learn integration slide, and use cross_val_score to evaluate your resulting model with 10-fold cross-validation.

**If you're using tensorflow-gpu on a Windows machine** by the way, you probably *do* want to peek a little bit at my solution - if you run into memory allocation errors, there's a workaround there you can use.

Try out your code here:

In [17]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score



model = Sequential()
    #16 feature inputs (votes) going into an 32-unit layer
model.add(Dense(32, input_dim=16, kernel_initializer='normal', activation='relu'))
    # Another hidden layer of 16 units
model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (Democrat or Republican political party)
model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


# Wrap our Keras model in an estimator compatible with scikit_learn
estimator =  model.fit(
    all_features, all_classes, epochs=10, verbose=1
)




Epoch 1/10
8/8 [==============================] - 1s 4ms/step - loss: 0.6928 - accuracy: 0.6379
Epoch 2/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6912 - accuracy: 0.8190
Epoch 3/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6887 - accuracy: 0.8190
Epoch 4/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6843 - accuracy: 0.8491
Epoch 5/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6762 - accuracy: 0.8707
Epoch 6/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6636 - accuracy: 0.8750
Epoch 7/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6446 - accuracy: 0.8836
Epoch 8/10
8/8 [==============================] - 0s 4ms/step - loss: 0.6173 - accuracy: 0.8750
Epoch 9/10
8/8 [==============================] - 0s 4ms/step - loss: 0.5792 - accuracy: 0.8750
Epoch 10/10
8/8 [==============================] - 0s 4ms/step - loss: 0.5307 - accuracy: 0.8836


In [19]:
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=10)
cv_scores.mean()

Epoch 1/10
8/8 [==============================] - 0s 3ms/step - loss: 0.4815 - accuracy: 0.8836
Epoch 2/10
8/8 [==============================] - 0s 4ms/step - loss: 0.4324 - accuracy: 0.8836
Epoch 3/10
8/8 [==============================] - 0s 7ms/step - loss: 0.3904 - accuracy: 0.9009
Epoch 4/10
8/8 [==============================] - 0s 4ms/step - loss: 0.3548 - accuracy: 0.9009
Epoch 5/10
8/8 [==============================] - 0s 4ms/step - loss: 0.3243 - accuracy: 0.9095
Epoch 6/10
8/8 [==============================] - 0s 4ms/step - loss: 0.2962 - accuracy: 0.9138
Epoch 7/10
8/8 [==============================] - 0s 5ms/step - loss: 0.2715 - accuracy: 0.9224
Epoch 8/10
8/8 [==============================] - 0s 4ms/step - loss: 0.2500 - accuracy: 0.9267
Epoch 9/10
8/8 [==============================] - 0s 5ms/step - loss: 0.2304 - accuracy: 0.9267
Epoch 10/10
8/8 [==============================] - 0s 6ms/step - loss: 0.2133 - accuracy: 0.9353


TypeError: estimator should be an estimator implementing 'fit' method, <keras.src.callbacks.History object at 0x7a48674d7bb0> was passed

## My implementation is below

# No peeking!

![title](peek.jpg)

In [10]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score

def create_model():
    model = Sequential()
    #16 feature inputs (votes) going into an 32-unit layer
    model.add(Dense(32, input_dim=16, kernel_initializer='normal', activation='relu'))
    # Another hidden layer of 16 units
    model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (Democrat or Republican political party)
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model



# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=create_model, epochs=100, verbose=0)
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=10)
cv_scores.mean()

ModuleNotFoundError: No module named 'tensorflow.keras.wrappers'

94% without even trying too hard! Did you do better? Maybe more neurons, more layers, or Dropout layers would help even more.